movieController.go

```diff
@@ -37,13 +45,304 @@ func moviesByName(c *gin.Context) {
+				result := db.Save(&entry)
+				if result.Error != nil {
+					log("ERROR", "Failed to post deletion to postgres")
+					c.JSON(http.StatusBadRequest, Response{Success: false, Context: result.Error.Error()})
+				} else {
+					log("INFO", "Deleted user")
+					c.JSON(http.StatusOK, Response{Success: true, Context: "Deleted User"})
+				}
+			}
+		} else {
+			log("ERROR", "Attempted to delete an inactive account")
+			c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Please enter a valid user ID"})
+		}
+	}
+}
+
+// updateUser godoc
+// @Summary updateUser
+// @Description update a user with a User-Id and new username and password
+// @Tags users
+// @Param User-Id header string true "UserID" 
+// @Param UserData body Users true "User Data"
+// @Accept */*
+// @Produce json
+// @Router /updateUser [PATCH]
+func updateUser(c *gin.Context) {
+	checkTime()
+	header := c.Request.Header["User-Id"]
+	if len(header) == 0 {
+		log("ERROR", "Invalid User-Id given")
+		c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Please enter a valid user ID"})
+	} else {
+		body, err := c.GetRawData()
+		if err != nil {
+			log("ERROR", "Failed to read request body")
+			c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Failed to read request body"})
+		} else {
+			var entry User
+			result := db.Where("id = ?", string(header[0])).First(&entry)
+			if entry.Active {
+				if result.Error != nil {
+					log("ERROR", "Invalid User-Id given")
+					c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Please enter a valid user ID"})
+				} else {
+					var creds Users
+					json.Unmarshal(body, &creds)
+					fmt.Println(creds)
+					hash, err := bcrypt.GenerateFromPassword([]byte(creds.Password), bcrypt.DefaultCost)
+					if err != nil {
+						log("ERROR", "Failed to encrypt")
+						c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Failed to encrypt"})
+					} else {
+						entry.Username = creds.Username
+						entry.Password = string(hash)
+						entry.UpdatedAt = time.Now().String()
+						result := db.Save(&entry)
+						if result.Error != nil {
+							log("ERROR", "Could not post user info to postgres")
+							c.JSON(http.StatusBadRequest, Response{Success: false, Context: result.Error.Error()})
+						} else {
+							log("INFO", "Updated user info")
+							c.JSON(http.StatusOK, Response{Success: true, Context: "Updated info"})
+						}
+					}
+				}
+			} else {
+				log("WARNING", "Attempted to update inactive account")
+				c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Please enter a valid user ID"})
+			}
+
+		}
+	}
+}
+
+// finishGame godoc
+// @Summary finishGame
+// @Description update success rates of user and daily
+// @Tags users
+// @Param User-Id header string false "UserID" 
+// @Param UserData body Users true "User Data"
+// @Accept */*
+// @Produce json
+// @Router /finishGame [POST]
+func finishGame(c *gin.Context) {
```

Comment:
This function is huge, I've asked gpt to refactor it & the results looks cleaner with less nesting. Let me know what you think & maybe we can have other big functions with big nesting refactored. 


```
func checkUserStatus(header []string) (User, bool) {
	if len(header) == 0 || len(header[0]) == 0 {
		log("ERROR", "Invalid user-id entered")
		return User{}, false
	}
	
	var user User
	result := db.Where("id = ?", string(header[0])).First(&user)
	if result.Error != nil || !user.Active {
		log("ERROR", "Invalid user ID entered")
		return User{}, false
	}
	
	return user, true
}

func updateGameState(status GameStatus, user *User) bool {
	if status.Won {
		user.SolvedPuzzles++
		user.LastSolvedPuzzle = time.Now().String()
	} else {
		user.FailedPuzzles++
	}
	
	result := db.Save(user)
	if result.Error != nil {
		log("ERROR", "Failed to post info to postgres")
		return false
	}
	
	log("INFO", "Submitted daily win info")
	return true
}

func updateSelectionStats(status GameStatus) {
	var daily selections
	result := db.Last(&daily)
	if result.Error != nil {
		log("ERROR", "Failed to pull info from postgres")
		return
	}
	
	if status.Won {
		daily.NumCorrect++
	} else {
		daily.NumIncorrect++
	}
	
	result = db.Model(&daily).Where("date = ?", daily.Date).Updates(map[string]interface{} {
		"NumCorrect": daily.NumCorrect,
		"NumIncorrect": daily.NumIncorrect, 
	})
	
	if result.Error != nil {
		log("ERROR", "Failed to post info to postgres")
	} else {
		log("INFO", "Submitted daily info")
	}
}

func finishGame(c *gin.Context) {
	checkTime()

	body, err := c.GetRawData()
	if err != nil {
		log("ERROR", "Failed to read request body")
		c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Failed to read request body"})
		return
	}

	var status GameStatus
	json.Unmarshal(body, &status)

	header := c.Request.Header["User-Id"]
	user, valid := checkUserStatus(header)
	if !valid {
		c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Please enter a valid user ID"})
		return
	}

	if !updateGameState(status, &user) {
		c.JSON(http.StatusBadRequest, Response{Success: false, Context: "Failed to post info to postgres"})
		return
	}
	
	updateSelectionStats(status)
	c.JSON(http.StatusOK, Response{Success: true, Context: "Submitted info"})
}

```

Reply:
GPT's solution does not account for guest scores, the function is large but it does more than the proposed function does.